### DISCUSSION TITLE
Grouping Pages for Faster Frontend Web View

### DISCUSSION THREAD

**Discussion by acrovitic** [py-pdf/pypdf#3000] [Q&A]
Hello,

I'm trying to group pages together within a PDFs internal structure, thus allowing my frontend to load pages from larger PDFs faster. When I talk about grouping together in this context, I mean the kind of grouping you'd find if inspecting a PDFs internal structure via preflight. For example, given a 25 page document with groups of 5 pages, you'd see 5 nodes in the internal structure under /Kids. 

I was hoping there was a more direct way to group pages as I need, but I didn't see anything in the documentation that could do this. I've also searched Issues and Discussion and I didn't find anything seemingly related to this problem I'm facing. If I missed a post relating to this, I apologize as it's likely that the appropriate terminology for what I am attempting is unfamiliar to me. 

Since I couldn't find any previous posts for this, I resorted to adjusting the internal structure of the PDF directly to group the pages. When I open the PDF in Adobe Acrobat, all I get are blank pages (following the example above, I only see 25 blank pages). However, the file size is as expected, using preflight to inspect the internal structure shows that the pages are grouped as expected, and when I open the file with a python script to retrieve content from the document, I am able to do so without issue. Sadly, I really need the content to be viewable via Adobe Acrobat for frontend. Any assistance is appreciated. My current code is below.

```
reader = PdfReader(input_path)
writer = PdfWriter()

root = reader.trailer["/Root"]
pages = root["/Pages"]
original_kids = pages["/Kids"]
group_size = 5 # 5 pages per group

for page in reader.pages: # copy pages to writer to preserve content
    writer.add_page(page)

grouped_kids = []
for i in range(0, len(original_kids), group_size):
    new_node = DictionaryObject() # make intermediate /Pages node
    new_kids = ArrayObject(original_kids[i : i + group_size]) # make subset of kids and add to grouping node
    new_node.update({
        NameObject("/Type"): NameObject("/Pages"),
        NameObject("/Kids"): new_kids,
        NameObject("/Count"): NumberObject(len(new_kids)),
    })
    new_node_ref = writer._add_object(new_node) # add grouping node to writer
    grouped_kids.append(new_node_ref)

pages[NameObject("/Kids")] = ArrayObject(grouped_kids) # update existing root with newly grouped Kids
pages[NameObject("/Count")] = NumberObject(len(original_kids))  
updated_pages_ref = writer._add_object(pages) # add updates /Pages to writer
root[NameObject("/Pages")] = updated_pages_ref # update root with updated /Pages
writer._root_object.update(root)

with open(output_path, "wb") as output_file: # export
    writer.write(output_file)
```

**Comment by stefan6419846** [comment]
Are you sure this is the correct solution for speeding up loading?

For the Adobe Reader part, we have a comment inside the code which states that your approach will not work there, as you have discovered yourself as well: https://github.com/py-pdf/pypdf/blob/27edc0612522689e44c1f9133c99bc221f34dbe7/pypdf/_writer.py#L476-L479
